### PR TITLE
Feature/issue 124

### LINKED ISSUE #124: PHP >= 7.4 tasks
Use PHP >= 7 features for library source code

For generated classes:
- [x] add `declare(strict_types=1)`: need to be thought before any implementation
- [x] add parameters typehint
- [x] add methods typehint
- [x] add a manifest about the generated classes
- [x] create the new major version

### PR DESCRIPTION
**Pull request by mikaelcom**
Implementations required to fix #124 and #179

### CODE REVIEW

**Comment by mikaelcom** [comment]
Anyone who has time to review the changes is welcome to add code-review comments :sweat_smile:

**Comment by mikaelcom** [comment]
Future enhancements are planned and listed in https://github.com/WsdlToPhp/PackageGenerator/issues/233, feel free to add more tasks :wink:

**Comment by mikaelcom** [comment]
Phar is available at https://github.com/WsdlToPhp/PackageGenerator/releases/tag/feature%2Fissue-124 and docker image is available at https://hub.docker.com/r/mikaelcom/wsdltophp

File: src/ConfigurationReader/GeneratorOptions.php [hunk 1]
```diff
@@ -1,614 +1,723 @@
 <?php
 
+declare(strict_types=1);
+
 namespace WsdlToPhp\PackageGenerator\ConfigurationReader;
 
-class GeneratorOptions extends AbstractYamlReader implements \JsonSerializable
+use InvalidArgumentException;
+
+final class GeneratorOptions extends AbstractYamlReader implements \JsonSerializable
```
Comment (mikaelcom) [inline src/ConfigurationReader/GeneratorOptions.php:9]: could use an import instead of `\JsonSerializable`

File: MANIFEST.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  - Read the [soapclient option](https://github.com/WsdlToPhp/PackageGenerator/wiki/Options#soapclient) for more information.
+- Classes contain custom method(s) in order to define the Soap Header(s) based on what is defined in the WSDL.
+  - Methods are named `setSoapHeader{SoapHeaderName}({SoapHeaderType} ${SoapHeaderName}, string $namespace = '{soapHeadernamespace}'}, bool $mustUnderstand = false, ?string $actor = null)`.
+- Classes contain one method per Soap operation **named after the operation name** all with respecting the **PHP method naming constraints**. The method returns:
+  - `false`:
+    - This means there is an error with the request or the Soap Server that has triggered a SoapFault (be sure that `WsdlToPhp\PackageBase\AbstractSoapClientBase::WSDL_TRACE`=`true`).
+    - Use `getLastError(string $methodName): ?SoapFault` to get the [SoapFault](https://www.php.net/manual/en/class.soapfault.php) thrown when the method cas called.
+    - `$methodName` has the form `{fully qualified classname}::{__FUNCION__}`.
```
Comment (mikaelcom) [inline MANIFEST.md]: typo `__FUNCION__` should be `__FUNCTION__`
```suggestion
    - `$methodName` has the form `{fully qualified classname}::{__FUNCTION__}`.
```

**Review by mikaelcom — COMMENTED** [review]
done

**Comment by faithfulman18** [comment]
Will these changes be compatible with PHP 8 as well as 7.4 ?  And will they be breaking changes that won't be backwards compatible?   Thanks.

**Comment by mikaelcom** [comment]
> Will these changes be compatible with PHP 8 as well as 7.4 ? And will they be breaking changes that won't be backwards compatible? Thanks.

The changes are compatible from PHP >= 7.4, so no more compatible to PHP < 7.4 both for the generator code and the generated code.

You can review the changes in the [UPGRADE-4.0](https://github.com/WsdlToPhp/PackageGenerator/blob/feature/issue-124/UPGRADE-4.0.md) and the [4.0.0 section of the CHANGELOG](https://github.com/WsdlToPhp/PackageGenerator/blob/feature/issue-124/CHANGELOG.md#400---2021-xx-xx)

**Comment by sprankhub** [comment]
Thanks a lot for all your work, @mikaelcom! I did not check the code, but tried the phar. This is what I noticed:

1. When using this new version, which includes strict types, I think the (simple) validation does not make much sense any more. I know I can disable it in the options, but still, I think it should not be added at all any more. I am talking about checking for `string` even though the parameter type already is `string`. Having validation for e.g. types inside arrays is still reasonable.
2. In the generated `ClassMap`, it would be much nicer to use `MyType::class` instead of `'\\MyType'`.

**Comment by mikaelcom** [comment]
Thanks @sprankhub, I really appreciate your feedback.

As there is always improvements to be made, I decided to create a new issue, https://github.com/WsdlToPhp/PackageGenerator/issues/233, gathering what would come next allowing to release the new version.

So feel free to add more things in this new issue especially your second comment as I already listed the first one.

**Pull request merged**